### ISSUE TITLE
Timezone support in Recurrence rules

### ISSUE THREAD

**Issue by tejasmanohar** [node-schedule/node-schedule#217]

**Comment by tejasmanohar** [comment]
As per implementation, using moment-timezone is probably our best bet, but I'm open to any other options. I'd think the API should look something like a recurrence rule accepting the option of `timezone`

i.e. `rule.timezone = 'America/Los_Angeles'`

**Comment by mrmayfield** [comment]
We may want to consider allowing another parameter with the same moment-timezone API representation of timezone for date-based scheduling,
for consistency's sake, especially if we're already adding it as a module dependency.

**Comment by tejasmanohar** [comment]
@mrmayfield I have no objections against it- just think having this feature in recurrence rules is of higher demand / need to users of the module. How would that look from an API perspective?

**Comment by tejasmanohar** [comment]
[time](https://github.com/TooTallNate/node-time) module is off-limits due to no windows support- @santigimeno integrated this earlier but i prematurely merged it and we reverted it in a major version update due to this issue. [timezone](https://github.com/bigeasy/timezone) and [moment-timezone](https://github.com/moment/moment-timezone/) are the good ones i've found so far... moment-timezone being the better one, imo.

**Comment by santigimeno** [comment]
I have some basic implementation using `timezone` but it's still a WIP and I don't know when / if it will be ready. I preferred it over `moment-timezone` because it's lighter. The major issue I find with any of these libraries is to know in which timezone the system is. 

As for the API, I think it should be independent of the type of `rule`, whether it's an `Object`, a `Date`, a `RecurrenceRule`, `String` ... As an example you can see what was my implementation using `node-time`: https://github.com/node-schedule/node-schedule/commit/ce8d7a6b27c0ead77b15d031de23a28be44cb096

**Comment by tejasmanohar** [comment]
Wow, I've been out of this for a while... kinda forgot about a lot of the changes that were on the `dev` branch now. Ok, I'll open up an issue about 1.0 and all the things we wanna add and just have others keep extending the list in comments

**Comment by tejasmanohar** [comment]
@santigimeno As per timezone integration, after looking at it more, I agree your approach is nicer than being dependent on the type of `rule`. @mrmayfield is working on timezone integration PR, as it's quite useful for us at AutoLotto. Would you mind sharing your WIP implementation that uses timezone? I don't mind between timezone and moment-timezone, as long as we stick to the standard timezone names like  'America/Los_Angeles' which it seems both libs do

**Comment by mrmayfield** [comment]
After some testing and experimentation, I've basically come up with two ways to implement timezone functionality in node-schedule that will work for *NIX/Microsoft platforms.

The first is to use [timezone-js](https://github.com/mde/timezone-js) which like the time module by TooTallNate extends the native JS Date object with timezone smarts. There are a couple problems with this module though, the first being that DST is not covered a'la the comments to [this answer](http://stackoverflow.com/a/10089167/3973137) on Stackoverflow. The second issue is where the module is a bit quirky, where it doesn't run out of the box but actually requires the timezone data to be downloaded and configured manually. What this would mean is adding the timezone data to the node-schedule module source, which is less than ideal.

The other option is a bit of a workaround. Tejas and I have determined that you can't use the local system's Date object's timezone to implement timezone functionality, even if you move the 'current time' forward or backwards by the timezone offset difference. This is because when the functionality builds out its next invocation date, it will always use the local system's JS Date object, which will end up being different from the desired time in the other timezone, and I can post a longer explanation if that needs to be elucidated a bit.

Thus, the workaround involves 'intercepting' the rule hour and moving it backwards/forwards by the difference in timezone offsets using moment-timezone. There are some edge cases where the hour moves back another day (ie, it goes 'below' 0) that can be accounted for with conditional logic. A similar approach can be used for moving the date back from 1 back to 7 if we are moving back over the beginning of a new week.

This approach is not as ideal as using a drop-in replacement for the JS Date object as TooTallNate's time module and node-timezone do, but for reasons listed above I think we'll probably find it a more desirable approach. Let me know what you guys think about the approach especially any concerns / better suggestions, otherwise I'd be happy to get to work on introducing that logic to work for Cron/non-cron scheduling & recurrence.

**Comment by tejasmanohar** [comment]
> This is because when the functionality builds out its next invocation date, it will always use the local system's JS Date object, which will end up being different from the desired time in the other timezone, and I can post a longer explanation if that needs to be elucidated a bit.

I think it's important to post the longer explanation with console.log's / date changes over time in the codebase for this one. And, _why_ this works for node-time (because it has timezone in the extended JS Date object) etc.

**Comment by joonhocho** [comment]
+1

**Comment by mj1856** [comment]
Hi everyone.  I just learned of this project, but I'd be glad to help, or at least give some advice.  I'm one of the maintainers of moment and moment-timezone, among other date/time related things on the interwebs...

I'll give a more thorough review over the code soon, but just wanted to say a couple of things quickly:

1. Timezone-js was a good attempt, but the [current implementation is buggy](https://github.com/mde/timezone-js/issues), especially near DST transitions.   There are a few other libraries [listed here](http://stackoverflow.com/a/15171030/634824). I'm admittedly biased on moment, but can say with confidence that is currently the better solution, and will be maintained over the long run.

2. If you're concerned about choosing a hard dependency, you *could* consider a plug-in model, where you route all time zone conversions through a function that can be implemented however the developer wishes.  That would allow for moment-timezone, another library, or a custom solution.  The down side is that it leaves the wiring to the consumer of the library, which some might consider non-obvious behavior.

3. Consider that others have indeed taken a hard dependency on moment.  [FullCalendar](http://fullcalendar.io/) comes to mind.  The timezone features of FullCalendar light up when you include moment-timezone, but the wiring is automatic because moment is already a dependency.

4. If node-schedule is truly aiming to be a general-purposes scheduling engine, then time zone support is not just optional, but *essential* (IMHO).  I've written about this [many](http://stackoverflow.com/a/19627330/634824), [times](http://stackoverflow.com/a/19170823/634824), [before](http://stackoverflow.com/a/26368158/634824).

5. Besides time zones, it's important to think about edge cases near DST transitions.  What do you do with a daily job that falls into the spring-forward gap of time that doesn't exist?  What about when the fall-back transition local times that occur twice?  There are a few sensible strategies, such as the one I described [here](http://nodatime.org/unstable/userguide/migration-to-2.html) "in the "Lenient resolver changes" section at the bottom of the page  (sorry, no anchor to link directly).  (This is on Noda Time, another project I contribute to.)

6. With great power comes great responsibility, and for scheduling systems, that means staying on top of time zone updates.  Be sure there's a clear strategy for how to do this with regards to your system, and make it clear in the documentation that this is essential.

I'll add more after a review of the code.  Cheers!

**Comment by santigimeno** [comment]
Wow that's a lot to digest! For sure I'll get back to you when I get to read it all. Thanks a lot for taking the time to write this!
BTW, there's still no code to review :(
One quick question: would you consider this: http://stackoverflow.com/questions/13195999/daylight-savings-and-cron valid strategy for DST transitions?

**Comment by mj1856** [comment]
From that link:

> If time has moved forward, those jobs that would have run in the interval that has been skipped will be run immediately.  Conversely, if time has moved backward, care is taken to avoid running jobs twice.

So, if the job is scheduled at 02:30, and the hour from `[02:00, 03:00)` is skipped, then the job would execute at 03:00 by these rules.  While this is a strategy I've seen before, it's usually not the one people expect - which would be to run at 03:30.  This is is the same change mentioned in the Noda Time library.

For the fall-back, that paragraph doesn't state which instance the job runs - just that the job doesn't run twice.  My recommendation is that the *first* of the two instances is preferred in most cases.

**Comment by santigimeno** [comment]
I've started implementing timezone support in `cron-parser` using `moment-timezone`. If it lands, adding support to `node-schedule` should be quite straightforward. See https://github.com/harrisiirak/cron-parser/pull/56. _

**Comment by santigimeno** [comment]
With the code from that branch and just applying this patch:
```
--- a/lib/schedule.js
+++ b/lib/schedule.js
@@ -368,7 +368,7 @@ RecurrenceRule.prototype.nextInvocationDate = function(base) {
     break;
   }
 
-  return next;
+  return next.date.toDate();
 };
```
our tests pass

**Comment by santigimeno** [comment]
We would only need then implementing the API and the tests

**Comment by mj1856** [comment]
@santigimeno - I made some comments on your PR.

**Comment by hankim813** [comment]
@santigimeno @tejasmanohar any update on this?

**Comment by santigimeno** [comment]
@hankim813 not really. I haven't had time to implement it. But if someone wants to give it a try, it should not be that difficult now that `cron-parser` has timezone support.

**Comment by damianobarbati** [comment]
+1!

**Comment by javimudi** [comment]
Hi,

👍  here for timezone support.

I've forked your repo and I am trying to understand which is the best place to add it. 

Just upgraded to the latest cron-parser release, but a major test break is produced.

Any advice about which the proper place to include the support is?

Regards

**Comment by santigimeno** [comment]
@javimudi it's great you're giving this a try! Thanks

Take a look at this: https://github.com/santigimeno/node-schedule/tree/bump_cron. The tests pass with latest cron-parser.

I would add a new `timezone` or `tz` option in the same way `start` and `end` options were added. Take a look at this PR: https://github.com/node-schedule/node-schedule/pull/292.

Just let us know if you need further advice. Good luck!

**Comment by javimudi** [comment]
Hi, @santigimeno,

I just added a PR following your advices and including several changes to handle timezone support.

I'll be waiting for your comments.

**Comment by santigimeno** [comment]
Great! I'll take a look as soon as time permits. Thanks!

**Comment by damianobarbati** [comment]
wow, got back here after 1 year for the same reason. still no support for timezone?

**Comment by damianobarbati** [comment]
@santigimeno did you had time to look into @javimudi PR? Timezone would be very useful indeed.

**Comment by zlatinejc** [comment]
they work if you specify the tz parameter when passing the ```spec``` object. It's mainly due to cron-parser supoort - the parsing is done via momemt / moment-timezone lib.

**Comment by Tethik** [comment]
I had to dig a bit through cron-parser and the source of this library, but this functionality does exist at least for cron-like jobs.

Example that works for me:
```js
schedule.scheduleJob(
    { rule: "21    17    *    *    1-5", tz: "Europe/Berlin" }, 
   () => console.log("hallo")
)
```

**Comment by RamParameswaran** [comment]
> I had to dig a bit through cron-parser and the source of this library, but this functionality does exist at least for cron-like jobs.
> ```

Agreed. I too had to dig through the source code to understand this functionality. 

Definitely a good case for a Documentation PR! 😉

**Comment by MarcGodard** [comment]
Sorry to initiate this after so long. However, I use cron strings, but want to pass a timezone. I just want to set everything to use UTC, therefore I can move the server's location / timezone and the cronstring would still be valid.

What would be the best way to do this?
